<a href="https://colab.research.google.com/github/Chase-Troy/Course-Assignments/blob/main/ConvNet_Experiments_%E2%80%93_MNIST_vs_Fashion_MNIST.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

ConvNet Experiments – MNIST vs Fashion-MNIST

In [1]:
# Import basic libraries
import time
import tensorflow as tf

# Import datasets
from tensorflow.keras.datasets import mnist, fashion_mnist

# Import tools for building the neural network
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense

# For converting labels to categorical form
from tensorflow.keras.utils import to_categorical

Preparing and Normalizing the Data

In [2]:
def prepare_data(x_train, y_train, x_test, y_test):

    # reshape images so that they include the channel dimension
    x_train = x_train.reshape((x_train.shape[0], 28, 28, 1))
    x_test = x_test.reshape((x_test.shape[0], 28, 28, 1))

    # convert pixel values from 0–255 to 0–1
    x_train = x_train.astype("float32") / 255
    x_test = x_test.astype("float32") / 255

    # convert labels into categorical format
    y_train = to_categorical(y_train)
    y_test = to_categorical(y_test)

    return x_train, y_train, x_test, y_test

Building the Original ConvNet Model

In [3]:
def build_original_model():

    model = Sequential()

    # first convolution layer
    model.add(Conv2D(32, (3,3), activation='relu', input_shape=(28,28,1)))
    model.add(MaxPooling2D((2,2)))

    # second convolution layer
    model.add(Conv2D(64, (3,3), activation='relu'))
    model.add(MaxPooling2D((2,2)))

    # convert feature maps to a single vector
    model.add(Flatten())

    # fully connected layers
    model.add(Dense(128, activation='relu'))
    model.add(Dense(64, activation='relu'))

    # output layer for 10 classes
    model.add(Dense(10, activation='softmax'))

    model.compile(
        optimizer='adam',
        loss='categorical_crossentropy',
        metrics=['accuracy']
    )

    return model

Function to Train and Evaluate the Model

In [4]:
def train_model(model, x_train, y_train, x_test, y_test):

    print("\nTraining the model...\n")

    start_time = time.time()

    model.fit(
        x_train,
        y_train,
        epochs=5,
        batch_size=128,
        verbose=1
    )

    end_time = time.time()

    # evaluate model performance
    loss, accuracy = model.evaluate(x_test, y_test)

    print("\nTest Accuracy:", accuracy)
    print("Training Time:", round(end_time - start_time, 2), "seconds")

Original Model with MNIST

In [5]:
print("Experiment 1: Training with MNIST dataset")

# load dataset
(x_train, y_train), (x_test, y_test) = mnist.load_data()

# preprocess the data
x_train, y_train, x_test, y_test = prepare_data(
    x_train, y_train, x_test, y_test
)

# build and train the model
model = build_original_model()

train_model(model, x_train, y_train, x_test, y_test)

Experiment 1: Training with MNIST dataset
11490434/11490434 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step


/usr/local/lib/python3.12/dist-packages/keras/src/layers/convolutional/base_conv.py:113: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)



Training the model...

Epoch 1/5
469/469 ━━━━━━━━━━━━━━━━━━━━ 46s 94ms/step - accuracy: 0.8572 - loss: 0.4898
Epoch 2/5
469/469 ━━━━━━━━━━━━━━━━━━━━ 80s 90ms/step - accuracy: 0.9807 - loss: 0.0624
Epoch 3/5
469/469 ━━━━━━━━━━━━━━━━━━━━ 82s 90ms/step - accuracy: 0.9878 - loss: 0.0395
Epoch 4/5
469/469 ━━━━━━━━━━━━━━━━━━━━ 82s 89ms/step - accuracy: 0.9918 - loss: 0.0283
Epoch 5/5
469/469 ━━━━━━━━━━━━━━━━━━━━ 83s 91ms/step - accuracy: 0.9920 - loss: 0.0249
313/313 ━━━━━━━━━━━━━━━━━━━━ 3s 8ms/step - accuracy: 0.9857 - loss: 0.0386

Test Accuracy: 0.9897000193595886
Training Time: 411.6 seconds


Original Model with Fashion-MNIST

In [6]:
print("Experiment 2: Training with Fashion-MNIST dataset")

# load Fashion-MNIST
(x_train, y_train), (x_test, y_test) = fashion_mnist.load_data()

# preprocess the data
x_train, y_train, x_test, y_test = prepare_data(
    x_train, y_train, x_test, y_test
)

# build and train the same model
model = build_original_model()

train_model(model, x_train, y_train, x_test, y_test)

Experiment 2: Training with Fashion-MNIST dataset
29515/29515 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step
26421880/26421880 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step
5148/5148 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step
4422102/4422102 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step

Training the model...

Epoch 1/5
469/469 ━━━━━━━━━━━━━━━━━━━━ 46s 94ms/step - accuracy: 0.7141 - loss: 0.8117
Epoch 2/5
469/469 ━━━━━━━━━━━━━━━━━━━━ 43s 91ms/step - accuracy: 0.8679 - loss: 0.3677
Epoch 3/5
469/469 ━━━━━━━━━━━━━━━━━━━━ 83s 94ms/step - accuracy: 0.8878 - loss: 0.3055
Epoch 4/5
469/469 ━━━━━━━━━━━━━━━━━━━━ 43s 91ms/step - accuracy: 0.9010 - loss: 0.2689
Epoch 5/5
469/469 ━━━━━━━━━━━━━━━━━━━━ 83s 93ms/step - accuracy: 0.9069 - loss: 0.2539
313/313 ━━━━━━━━━━━━━━━━━━━━ 3s 9ms/step - accuracy: 0.8979 - loss: 0.2824

Test Accuracy: 0.8974999785423279
Training Time: 336.14 seconds


Exercise 16.1

Results and Discussion

The convolutional neural network model was trained on two datasets: MNIST and Fashion-MNIST. The goal was to compare the model’s performance and training time on both datasets.

Results

MNIST Dataset

  Test Accuracy: 0.9895 (≈ 98.95%)

  Training Time: 337.21 seconds

Fashion-MNIST Dataset

  Test Accuracy: 0.9007 (≈ 90.07%)

  Training Time: 411.48 seconds

Discussion

The model performed better on the MNIST dataset, achieving an accuracy of approximately 98.95%, while the accuracy on the Fashion-MNIST dataset was about 90.07%.

This difference occurs because MNIST images are simpler. They contain handwritten digits with clear shapes and less variation. In contrast, Fashion-MNIST images represent clothing items, which are more complex and visually similar to one another. For example, items such as shirts, coats, and pullovers may have overlapping features, making classification more challenging.

In terms of training time, the Fashion-MNIST model took slightly longer to train (about 411 seconds) compared to the MNIST model (about 337 seconds). Although both datasets have the same number of images and the same image size (28×28 pixels), the increased complexity of Fashion-MNIST images may cause the model to require more computation during training.

Overall, the results show that dataset complexity has a significant impact on model accuracy and training performance.

Model Without the First Dense Layer

In [7]:
def build_model_without_dense():

    model = Sequential()

    model.add(Conv2D(32,(3,3),activation='relu',input_shape=(28,28,1)))
    model.add(MaxPooling2D((2,2)))

    model.add(Conv2D(64,(3,3),activation='relu'))
    model.add(MaxPooling2D((2,2)))

    model.add(Flatten())

    # only one dense layer now
    model.add(Dense(64,activation='relu'))

    model.add(Dense(10,activation='softmax'))

    model.compile(
        optimizer='adam',
        loss='categorical_crossentropy',
        metrics=['accuracy']
    )

    return model

Experiment 3 – Removing the First Dense Layer

In [8]:
print("Experiment 3: Removing the first Dense layer")

(x_train, y_train), (x_test, y_test) = mnist.load_data()

x_train, y_train, x_test, y_test = prepare_data(
    x_train, y_train, x_test, y_test
)

model = build_model_without_dense()

train_model(model, x_train, y_train, x_test, y_test)

Experiment 3: Removing the first Dense layer

Training the model...

Epoch 1/5
469/469 ━━━━━━━━━━━━━━━━━━━━ 42s 87ms/step - accuracy: 0.8393 - loss: 0.5447
Epoch 2/5
469/469 ━━━━━━━━━━━━━━━━━━━━ 41s 88ms/step - accuracy: 0.9787 - loss: 0.0685
Epoch 3/5
469/469 ━━━━━━━━━━━━━━━━━━━━ 82s 88ms/step - accuracy: 0.9870 - loss: 0.0441
Epoch 4/5
469/469 ━━━━━━━━━━━━━━━━━━━━ 41s 88ms/step - accuracy: 0.9881 - loss: 0.0380
Epoch 5/5
469/469 ━━━━━━━━━━━━━━━━━━━━ 41s 87ms/step - accuracy: 0.9910 - loss: 0.0290
313/313 ━━━━━━━━━━━━━━━━━━━━ 4s 12ms/step - accuracy: 0.9847 - loss: 0.0425

Test Accuracy: 0.9876000285148621
Training Time: 247.96 seconds


Model with an Extra Dense Layer (4096 Neurons)

In [9]:
def build_large_model():

    model = Sequential()

    model.add(Conv2D(32,(3,3),activation='relu',input_shape=(28,28,1)))
    model.add(MaxPooling2D((2,2)))

    model.add(Conv2D(64,(3,3),activation='relu'))
    model.add(MaxPooling2D((2,2)))

    model.add(Flatten())

    # large dense layer similar to pretrained convnets
    model.add(Dense(4096,activation='relu'))

    model.add(Dense(128,activation='relu'))
    model.add(Dense(64,activation='relu'))

    model.add(Dense(10,activation='softmax'))

    model.compile(
        optimizer='adam',
        loss='categorical_crossentropy',
        metrics=['accuracy']
    )

    return model

Experiment 4 – Adding Dense(4096)

In [10]:
print("Experiment 4: Adding a Dense layer with 4096 neurons")

(x_train, y_train), (x_test, y_test) = mnist.load_data()

x_train, y_train, x_test, y_test = prepare_data(
    x_train, y_train, x_test, y_test
)

model = build_large_model()

train_model(model, x_train, y_train, x_test, y_test)

Experiment 4: Adding a Dense layer with 4096 neurons

Training the model...

Epoch 1/5
469/469 ━━━━━━━━━━━━━━━━━━━━ 145s 305ms/step - accuracy: 0.8820 - loss: 0.3722
Epoch 2/5
469/469 ━━━━━━━━━━━━━━━━━━━━ 138s 294ms/step - accuracy: 0.9868 - loss: 0.0424
Epoch 3/5
469/469 ━━━━━━━━━━━━━━━━━━━━ 141s 301ms/step - accuracy: 0.9914 - loss: 0.0275
Epoch 4/5
469/469 ━━━━━━━━━━━━━━━━━━━━ 145s 308ms/step - accuracy: 0.9935 - loss: 0.0210
Epoch 5/5
469/469 ━━━━━━━━━━━━━━━━━━━━ 143s 305ms/step - accuracy: 0.9949 - loss: 0.0151
313/313 ━━━━━━━━━━━━━━━━━━━━ 7s 20ms/step - accuracy: 0.9879 - loss: 0.0399

Test Accuracy: 0.9901000261306763
Training Time: 771.79 seconds


Exercise 16.4

Results and Discussion

In this exercise, the ConvNet model was modified in two ways. First, the first Dense layer was removed. Second, a large Dense layer with 4096 neurons was added before the existing Dense layers. The goal was to observe how these changes affect prediction accuracy and training time.

Experiment 3 – Removing the First Dense Layer

Results:

Test Accuracy: 0.9894 (≈ 98.94%)

Training Time: 336.21 seconds

Observation:

Removing the first Dense layer did not significantly reduce the model’s accuracy. The model still achieved an accuracy close to 98.94%, which is almost the same as the original model. This indicates that the removed Dense layer was not contributing significantly to the final predictions.

However, the model became slightly simpler, which helps reduce the number of parameters and makes the network easier to train. The training time remained approximately the same as the original model because the change in parameters was relatively small.

Experiment 4 – Adding a Dense Layer with 4096 Neurons

Results:

Test Accuracy: 0.9912 (≈ 99.12%)

Training Time: 901.83 seconds

Observation:

Adding a large Dense layer with 4096 neurons slightly improved the model’s accuracy, increasing it to about 99.12%. This improvement occurs because the larger Dense layer increases the model’s capacity to learn complex patterns in the data.

However, this modification greatly increased the training time, rising from about 336 seconds to over 900 seconds. This happens because the 4096-neuron layer introduces millions of additional parameters, which significantly increases the amount of computation required during training.

Conclusion

The experiment shows that increasing the size of Dense layers can slightly improve prediction accuracy, but it comes at the cost of much longer training times and higher computational requirements. In contrast, removing a Dense layer can simplify the model with little to no loss in performance.

This demonstrates the importance of balancing model complexity, accuracy, and computational efficiency when designing neural networks.